src/Cache/CacheEngine.php

```diff
@@ -94,20 +98,25 @@ protected function ensureValidKey($key)
     }
 
     /**
-     * Ensure the validity of the given cache keys.
+     * Ensure the validity of the argument type and cache keys.
      *
-     * @param mixed $keys The keys to check.
+     * @param iterable $iterable The iterable to check.
+     * @param string $check Whether to check keys or values.
      * @return void
-     * @throws \Cake\Cache\InvalidArgumentException When the keys are not valid.
+     * @throws \Cake\Cache\InvalidArgumentException
      */
-    protected function ensureValidKeys($keys)
+    protected function ensureValidType($iterable, string $check = self::CHECK_VALUE)
     {
-        if (!is_array($keys) && !($keys instanceof \Traversable)) {
-            throw new InvalidArgumentException('A cache key set must be either an array or a Traversable.');
+        if (!is_iterable($iterable)) {
+            throw new InvalidArgumentException('A cache set / keys set must be either an array or a Traversable.');
```

Comment:
This error message is confusing to me. Can we use `$check` to make it easier for the user to know which mode failed?